File: offer/30/JPhoebe.py
# -*- coding:utf-8 -*-
class Solution:
    def FindGreatestSumOfSubArray(self, array):
        res = array[0]
        for i in range(0, len(array)):
            max = array[i]
            temp_max = max
            for j in range(i+1, len(array)):
                temp_max = temp_max + array[j]
                if temp_max > max:
                    max = temp_max
            if max > res:
                res = max
        print(res)
        return res

Solution().findgreatestsumofsubarray([-2,-8,-1,-5,-9])
# Solution().findgreatestsumofsubarray([1,-2,3,10,-4,7,2,-5])
# Solution().findgreatestsumofsubarray([-1, 2, -2, -3, -4])

File: offer/67/jphoebe.java
public class Solution {
    public int cutRope(int target) {
        if (target < 0) {
            return 0;
        } else if (target == 2) {
            return 1;
        } else if (target == 3) {
            return 2;
        } else{
            int list[] = new int[target + 1];
            list[0] = 0;
            list[1] = 1;
            list[2] = 2;
            list[3] = 3;
            for (int i = 4; i <= target; i++) {
                int max = 0;
                for (int j = 1; j <= i / 2; j++) {
                    int temp = list[j] * list[i - j];
                    if (max < temp) {
                        max = temp;
                    }
                }
                list[i] = max;
            }
            return list[target];
        }
    }
}
